build(docs-infra): simplify the docs examples update workflow and docs

gkalpak opened the pull request:
This commit simplifies/improves the workflow (and associated docs) for updating boilerplate for docs examples projects.
See individual commits for more details.

mary-poppins commented:
You can preview b235293 at https://pr38992-b235293.ngbuilds.io/.

petebacondarwin commented on aio/tools/examples/UPDATING.md:
```diff
@@ -0,0 +1,55 @@
+  > In order for `ng update` to work, there must be a `node_modules/` directory with installed dependencies inside the [shared/boilerplate/cli/](./shared/boilerplate/cli) directory.
+  > This `node_modules/` directory is only needed during the update operation and must be subsequently removed to avoid its being copied into all CLI-based examples.
+
+- The previous command made any necessary changes to boilerplate files inside the `cli/` folder, but the same changes need to be applied to the other CLI-based boilerplate folder.
```
```suggestion
- The previous command made any necessary changes to boilerplate files inside the `cli/` folder, but the same changes need to be applied to the other CLI-based boilerplate folders.
```

petebacondarwin commented on aio/tools/examples/example-boilerplate.spec.js:
```diff
@@ -317,6 +335,33 @@ describe('example-boilerplate tool', () => {
+
+        // Skip `sub-dir-1` and all its contents.
+      ]);
+    });
```
Nice test!

petebacondarwin commented on aio/tools/examples/shared/sync-boilerplate-dependencies.js:
```diff
@@ -0,0 +1,67 @@
+#!/usr/bin/env node
```
I don't think this line is needed if you are running use `node sync-boilerplate-dependencies`.

gkalpak replied:
No, it's not, but this allows running it without `node` (I mean without explicitly typing `node`) and we have it in many (most?) similar scripts. So, since it doesn't hurt keeping it...I think we should keep it :smiley:

petebacondarwin commented on aio/tools/examples/shared/sync-boilerplate-dependencies.js:
```diff
@@ -0,0 +1,67 @@
+}
+
+function loadJsonFile(filePath) {
+  return JSON.parse(fs.readFileSync(filePath, 'utf8'));
```
Can also use `require(filePath)` here?

gkalpak replied:
You could, but you have to ensure that (now or in the future) you are not loading a file again, after having updated it on disk. Because `require(...)` will cache the result after the first load, so it won't see any updates.

Since in the purpose of this script is to both load _and_ update JSON files on disk, I think it is safer to use `readFileSync()`.

petebacondarwin commented on aio/tools/examples/shared/sync-boilerplate-dependencies.js:
```diff
@@ -0,0 +1,67 @@
+function syncDependencies(boilerplatePkgJsonPath) {
+  console.log(`Syncing '${path.relative(__dirname, boilerplatePkgJsonPath)}'...`);
+
+  const boilerplatePkgJson = loadJsonFile(boilerplatePkgJsonPath);
```
NIT: I would change this function to make it only do one thing.

Currently it loads the file and then changes the object. But it doesn't save it. 

So either it should do both the loading and saving, or it should do neither and leave that to the caller.

gkalpak replied:
Originally, I had it do the updating only, but then I needed the path for the error message. So, I thought why not do the loading here anyway. I'll move the saving into this function as well, since there is little benefit having it separate atm.

petebacondarwin reviewed (approved):
Nice tidy up. I left quite a lot of minor niggles feel free to take what works for you.

gkalpak committed:
docs(docs-infra): apply the one-sentence-per-line rule to Markdown files in `aio/tools/`

This commit updates the Markdown files inside the `aio/tools/` directory
to contain one sentence per line in order to be consistent with how
Markdown files are formatted in the rest of the repo.

gkalpak committed:
docs(docs-infra): update instructions for updating docs examples dependencies

gkalpak committed:
build(docs-infra): simplify update workflow for CLI-based docs examples boilerplate

When updating the boilerplate for CLI-based docs examples, one needed to
install dependencies inside the
`aio/tools/examples/shared/boilerplate/cli/` directory, which resulted
in a `node_modules/` directory and a `yarn.lock` file. These were not
supposed to be part of the boilerplate, so they had to be manually
removed after the boilerplate was updated.

This commit simplifies the workflow by allowing boilerplate files to be
ignored (both by git and the `example-boilerplate.js` script) via a
`.gitignore` file. This way, it is no longer necessary to manually
remove the unneeded directories/files.

gkalpak committed:
build(docs-infra): simplify updating dependencies in docs examples boilerplate

Previously, when updating the dependency versions in
`aio/tools/examples/shared/package.json` (which contains all
dependencies used in docs examples projects), one had to manually go
through all boilerplate directories and update the `package.json` files
with the same versions.

This commit simplifies this task by automating it via a Node.js script.

gkalpak commented on aio/tools/examples/README.md:
```diff
@@ -1,106 +1,151 @@
+As mentioned above, many of the documentation pages contain snippets extracted from real example applications.
+To achieve that, all those applications needs to contain some basic boilerplate, such as a `node_modules/` folder, a `package.json` file with scripts and dependencies, etc.
+
+To avoid having to maintain the boilerplate in each example, we use the [example-boilerplate-js](./example-boilerplate.js) script to provide a set of files that works across all the examples.
```
Fair enough :smiley: I'll update it to mention project types.

gkalpak commented on aio/tools/examples/README.md:
```diff
@@ -1,106 +1,151 @@
+- `common`: Contains files used in many examples.
+  (See the [next section](#example-config) for info on how to not include common files in certain examples.)
+- `viewengine/cli`: Additional configuration for running CLI-based examples with `ViewEngine` (the pre-Ivy compiler/renderer).
+  This applies to all CLI-based examples, such as `cli-ajs`, `elements`, `getting-started`, etc.
```
Yes (when run in ViewEngine mode).

Basically, we can choose whether to run the tests in the default, Ivy mode or in ViewEngine mode. (We run both on CI, on two different jobs.)

In Ivy mode, all these project types get the files from `boilerplate/cli/` plus the ones in the respective project-type directory (e.g. `boilerplate/cli-ajs/`, `boilerplate/elements/`, etc.). Similarly, in ViewEngine mode, these project types will additionally get `boilerplate/viewengine/cli/`.

For reference, here is the [source code][1].

[1]: https://github.com/angular/angular/blob/323be39297450fa5a4a70f800511722b4964dec2/aio/tools/examples/example-boilerplate.js#L47-L67

gkalpak commented on aio/tools/examples/README.md:
```diff
@@ -1,106 +1,151 @@
-### A node_modules to share
+### A `node_modules/` to share
 
 With all the boilerplate files in place, the only missing piece are the installed packages.
```
I went with the first suggestion 🤷‍♂️

gkalpak commented on aio/tools/examples/README.md:
```diff
@@ -1,106 +1,151 @@
+For that we have [shared/package.json](./shared/package.json), which contains **all** the packages needed to run any example type.
 
-After installing these dependencies, a `node_modules/` folder will be created at `/aio/tools/examples/shared/node_modules/`.
+After installing these dependencies, a [shared/node_modules/](./shared/node_modules) folder will be created.
```
What we want to say here is that the `shared/node_modules/` folder is created as a result of installing the dependencies. So, _after_ sounds a bit misleading to me.

I changed it to: _"Upon_ installing [...] a [...] folder _is_ created"_

gkalpak committed:
fixup! docs(docs-infra): update instructions for updating docs examples dependencies

gkalpak committed:
fixup! build(docs-infra): simplify updating dependencies in docs examples boilerplate

mary-poppins commented:
You can preview 0c96129 at https://pr38992-0c96129.ngbuilds.io/.
